src/components/mc/cuda/mc_cuda.c

```diff
@@ -183,9 +338,120 @@ static ucc_status_t ucc_mc_cuda_mem_query(const void *ptr,
     return UCC_OK;
 }
 
+
+ucc_status_t ucc_ee_cuda_task_post(void *ee_stream, void **ee_req)
+{
+    ucc_mc_cuda_stream_request_t *req;
+    ucc_mc_cuda_event_t *cuda_event;
+    ucc_status_t status;
+
+    req = ucs_mpool_get(&ucc_mc_cuda.strm_reqs);
+    ucc_assert(req);
+    req->status = UCC_MC_CUDA_TASK_POSTED;
+    req->stream = (cudaStream_t)ee_stream;
+
+    if (ucc_mc_cuda.task_strm_type == UCC_MC_CUDA_USER_STREAM) {
+        status = ucc_mc_cuda.post_strm_task(req->dev_status, req->stream);
+        if (status != UCC_OK) {
+            goto free_req;
+        }
+    } else {
+        cuda_event = ucs_mpool_get(&ucc_mc_cuda.events);
+        ucc_assert(cuda_event);
+        CUDACHECK(cudaEventRecord(cuda_event->event, req->stream));
+        CUDACHECK(cudaStreamWaitEvent(ucc_mc_cuda.stream, cuda_event->event, 0));
+        status = ucc_mc_cuda.post_strm_task(req->dev_status, ucc_mc_cuda.stream);
+        if (status != UCC_OK) {
+            goto free_event;
+        }
+        CUDACHECK(cudaEventRecord(cuda_event->event, ucc_mc_cuda.stream));
+        CUDACHECK(cudaStreamWaitEvent(req->stream, cuda_event->event, 0));
+        ucs_mpool_put(cuda_event);
+    }
+
+    *ee_req = (void *) req;
+
+    mc_info(&ucc_mc_cuda.super, "CUDA stream task posted on \"%s\" stream. req:%p",
+            task_stream_types[ucc_mc_cuda.task_strm_type], req);
+
+    return UCC_OK;
+
+free_event:
+    ucs_mpool_put(cuda_event);
+free_req:
+    ucs_mpool_put(req);
+    return status;
+}
+
+ucc_status_t ucc_ee_cuda_task_query(void *ee_req)
```

Comment:
why all interfaces with "void" req? Usually we would define ucc_ee_req_t "super" request (i guess some ucc_ee_status could be there) and components would extend it and use "ucc_derived_of". I know it is only a visual difference but it would be more consistent with the rest of the code.